product/uppercut/template.builder/TemplateDictionary.cs

```diff
@@ -10,13 +13,27 @@ public sealed class TemplateDictionary
             IDictionary<string, string> token_replacement_dictionary = new Dictionary<string, string>();
             token_replacement_dictionary.Add("${quote}", "\"");
 
-            foreach (Match name_value_match in get_matches(settings_text))
+            using (var text = new StringReader(settings_text))
+            using (var reader = XmlReader.Create(text, new XmlReaderSettings { IgnoreWhitespace = true }))
             {
-                string pair_key_text = get_name_value(name_value_match.ToString());
-                string pair_value_text = get_pair_value(name_value_match.ToString(), token_replacement_dictionary);
-                pair_key_text = @"${" + pair_key_text + @"}";
-                if (!token_replacement_dictionary.ContainsKey(pair_key_text))
-                    token_replacement_dictionary.Add(pair_key_text, pair_value_text);
+                while (reader.Read())
+                {
+                    switch (reader.NodeType)
+                    {
+                        case XmlNodeType.Element:
+                            if (reader.Name == "property")
+                            {
+                                var name = reader["name"];
+                                var value = reader["value"] ?? reader.ReadInnerXml();
+                                if (!string.IsNullOrEmpty(value))
+                                {
+                                    var replaced = replace_tokens_with_other_dictionary_values(value, token_replacement_dictionary);
+                                    token_replacement_dictionary.Add("${" + name + "}", replaced);
+                                }
+                            }
+                            break;
+                    }
+                }

```

Comment:
So what I'm seeing is that we are treating the entire document as an XML file instead of reading through a text file.